std/sumtype.d

```diff
@@ -0,0 +1,2504 @@
+    }
+
+    static foreach (tid, T; Types)
+    {
+        static if (isAssignableTo!T)
+        {
+            /**
+             * Assigns a value to a `SumType`.
+             *
+             * Assigning to a `SumType` is `@system` if any of the
+             * `SumType`'s members contain pointers or references, since
+             * those members may be reachable through external references,
+             * and overwriting them could therefore lead to memory
+             * corruption.
+             *
+             * An individual assignment can be `@trusted` if the caller can
+             * guarantee that there are no outstanding references to $(I any)
+             * of the `SumType`'s members when the assignment occurs.
+             */
+            ref SumType opAssign(T rhs)
+            {
+                import core.lifetime : forward;
+                import std.traits : hasIndirections, hasNested;
+                import std.meta : Or = templateOr;
+
+                enum mayContainPointers =
+                    anySatisfy!(Or!(hasIndirections, hasNested), Types);
+
+                static if (mayContainPointers)
+                {
+                    cast(void) () @system {}();
+                }
+
+                this.match!destroyIfOwner;
+
+                mixin("Storage newStorage = { ",
+                    Storage.memberName!T, ": forward!rhs",
+                " };");
+
+                storage = newStorage;
+                tag = tid;
+
+                return this;
+            }
+        }
+    }
+
+    static if (allSatisfy!(isAssignableTo, Types))
+    {
+        static if (allSatisfy!(isCopyable, Types))
+        {
+            /**
+             * Copies the value from another `SumType` into this one.
+             *
+             * See the value-assignment overload for details on `@safe`ty.
+             *
+             * Copy assignment is `@disable`d if any of `Types` is non-copyable.
+             */
+            ref SumType opAssign(ref SumType rhs)
+            {
+                rhs.match!((ref value) { this = value; });
+                return this;
+            }
+        }
+        else
+        {
+            @disable ref SumType opAssign(ref SumType rhs);
+        }
+
+        /**
+         * Moves the value from another `SumType` into this one.
+         *
+         * See the value-assignment overload for details on `@safe`ty.
+         */
+        ref SumType opAssign(SumType rhs)
+        {
+            import core.lifetime : move;
+
+            rhs.match!((ref value) { this = move(value); });
+            return this;
+        }
+    }
+
+    /**
+     * Compares two `SumType`s for equality.
+     *
+     * Two `SumType`s are equal if they are the same kind of `SumType`, they
+     * contain values of the same type, and those values are equal.
+     */
+    bool opEquals(this This, Rhs)(auto ref Rhs rhs)
+    if (is(CommonType!(This, Rhs)))
+    {
+        static if (is(This == Rhs))
```

Comment:
Is this not a problem if they differ in constness?

Reply:
That's what the `else` branch is for. The `unittest` labelled "Equality of differently-qualified SumTypes" covers it.